controllers/managedocs_controller.go

```diff
@@ -604,39 +649,87 @@ func (r *ManagedOCSReconciler) reconcileAlertmanagerConfigSecret() error {
-		} `yaml:"route"`
-		Receivers [1]struct {
-			Name             string `yaml:"name"`
-			PagerdutyConfigs [1]struct {
-				ServiceKey string `yaml:"service_key"`
-			} `yaml:"pagerduty_configs"`
-		} `yaml:"receivers"`
-	}{}
-
-	alertmanagerConfig.Route.GroupWait = "30s"
-	alertmanagerConfig.Route.GroupInterval = "5m"
-	alertmanagerConfig.Route.RepeatInterval = "4h"
-	alertmanagerConfig.Route.Receiver = "pagerduty"
-	alertmanagerConfig.Route.Routes[0].Match.Severity = "critical"
-	alertmanagerConfig.Route.Routes[0].Receiver = "pagerduty"
-	alertmanagerConfig.Receivers[0].Name = "pagerduty"
-	alertmanagerConfig.Receivers[0].PagerdutyConfigs[0].ServiceKey = pagerdutyServiceKey
-
-	return &alertmanagerConfig
+// Prometheus-operator v0.37.0 takes its alert manager configuration from a secret containing a yaml file.
+// This function produces that yaml file.
+// To sanitize the input the yaml is first represented as types in golang.
+func (r *ManagedOCSReconciler) generateAlertmanagerConfig(pagerdutyServiceKey string, dmsURL string) interface{} {
+
+	type webhookConfig struct {
+		URL string `yaml:"url,omitempty"`
+	}
+
+	type pagerdutyConfig struct {
+		ServiceKey string `yaml:"service_key,omitempty"`
+	}
+
+	type route struct {
+		Receiver       string            `yaml:"receiver,omitempty"`
+		Match          map[string]string `yaml:"match,omitempty"`
+		Routes         []*route          `yaml:"routes,omitempty"`
+		GroupWait      string            `yaml:"group_wait,omitempty"`
+		GroupInterval  string            `yaml:"group_interval,omitempty"`
+		RepeatInterval string            `yaml:"repeat_interval,omitempty"`
+	}
+
+	type receiver struct {
+		Name             string             `yaml:"name"`
+		PagerdutyConfigs []*pagerdutyConfig `yaml:"pagerduty_configs,omitempty"`
+		WebhookConfigs   []*webhookConfig   `yaml:"webhook_configs,omitempty"`
+	}
+
+	type alertmanagerConfig struct {
+		Route     *route      `yaml:"route,omitempty"`
+		Receivers []*receiver `yaml:"receivers,omitempty"`
+	}
+
+	amConfig := alertmanagerConfig{
+		Route: &route{
+			Receiver: "deadmanssnitch",
+			Routes: []*route{
+				&route{
+					Receiver:       "pagerduty",
+					GroupWait:      "30s",
+					GroupInterval:  "5m",
+					RepeatInterval: "4h",
+					Match: map[string]string{
+						"severity": "critical",
+					},
+				},
+				&route{
+					Receiver:       "deadmanssnitch",
+					GroupWait:      "5m",
+					GroupInterval:  "5m",
+					RepeatInterval: "5m",
+					Match: map[string]string{
+						"alertname": "DeadMansSnitch",
+					},
+				},
+			},
+		},
+		Receivers: []*receiver{
+			&receiver{
+				Name: "default-receiver",
+			},
+			&receiver{
+				Name: "pagerduty",
+				PagerdutyConfigs: []*pagerdutyConfig{
+					&pagerdutyConfig{
+						ServiceKey: pagerdutyServiceKey,
+					},
+				},
+			},
+			&receiver{
+				Name: "deadmanssnitch",
+				WebhookConfigs: []*webhookConfig{
+					&webhookConfig{
+						URL: dmsURL,
+					},
+				},
+			},
+		},
+	}
```

Comment:
What is the reason behind this refactor? I can see a couple of reasons why the change should be rejected:
1. It just creates much more code where it is not needed
2. It creates a list of named types that are used only once, which are not needed because we do not reuse them
3. It introduces pointer logic into the structures 
4. And the most important one: It undoes a conscious decision made by under contributor, just recently, without an explanation.

Reply:
This refactor was necessary for me to accomplish what this PR sets out to do. I struggled to add the needed fields to the unnamed struct that was originally there. I followed your suggestion and wound up with the following yaml:
```
route:                                                                                                                                             
  group_wait: 30s                                                                                                                                  
  group_interval: 5m                                                                                                                               
  repeat_interval: 4h                                                                                                                              
  receiver: pagerduty                                                                                                                              
  routes:                                                                                                                                          
  - receiver: pagerduty                                                                                                                            
    group_wait: 30s                                                                                                                                
    group_interval: 5m                                                                                                                             
    repeat_interval: 4h                                                                                                                            
    match:                                                                                                                                         
      severity: critical                                                                                                                           
  - receiver: deadmanssnitch                                                                                                                       
    group_wait: 5m                                                                                                                                 
    group_interval: 5m                                                                                                                             
    repeat_interval: 5m                                                                                                                            
    match:                                                                                                                                         
      alertname: deadmanssnitch
receivers:
- name: pagerduty
  pagerduty_configs:
  - service_key: FOO
  webhook_configs:
  - {}
- name: deadmanssnitch
  pagerduty_configs:
  - {}
  webhook_configs:
  - url: BAR
 ```
I didn't like that it lists the webhook_configs and pagerduty_configs even when there isn't one there, and couldn't figure out a way to change it. Instead of spending too much time on that I thought it was cleaner and easier to write it in the way that I did. Because of the points you made here, I tried running the operator with the code that generates the above yaml, but it does break the alertmanager with the following error:
```
level=error ts=2021-04-23T20:11:43.662Z caller=coordinator.go:124 component=configuration msg="Loading configuration file failed" file=/etc/alertmanager/config/alertmanager.yaml err="missing URL in webhook config"
``` 
If it needs to be changed to an unnamed struct, I'll need help to write it in a satisfactory way.

To speak to your 4th point (I was confused why it's last on the list if it's the most important), it was not my intention to undo something someone else did. The unnamed struct approach worked when the config was only for pagerduty, I just don't have the abilities to expand it and did what I could to get it to work.

Reply:
Ok, I understand the issue and the point takes. Next time, it will be helpful to add this kind of explanation to the PR description. 

Saying that I would prefer the more minimal change that does not introduce pointer logic and sticks more closely to the original contributor's solution (will not introduce empty arrays or structures): 
 ```suggestion
	type PagerDutyConfig struct {
	    ServiceKey string `yaml:"service_key"`
	}
	
	type WebhookConfig struct {
	    Url string `yaml:"url"`
	}

	alertmanagerConfig := struct {
		Route struct {
			Routes         [2]struct {
				GroupWait       string `yaml:"group_wait"`
				GroupInterval  string `yaml:"group_interval"`
				RepeatInterval string `yaml:"repeat_interval"`
				Receiver           string `yaml:"receiver"`
				Match struct {
					Severity string `yaml:"severity"`
				} `yaml:"match"`
			} `yaml:"routes"`
		} `yaml:"route"`
		Receivers [2]struct {
			Name                         string `yaml:"name"`
			PagerdutyConfigs    []PagerDutyConfig `yaml:"pagerduty_configs,omitempty"`
			WebhookConfigs 	 []WebhookConfig   `yaml:"webhooks_configs,omitempty"`
		} `yaml:"receivers"`
	}{}
	
	alertmanagerConfig.Route.Routes[0].GroupWait = "30s"
	alertmanagerConfig.Route.Routes[0].GroupInterval = "5m"
	alertmanagerConfig.Route.Routes[0].RepeatInterval = "4h"
	alertmanagerConfig.Route.Routes[0].Receiver = "pagerduty"
	alertmanagerConfig.Route.Routes[0].Match.Severity = "critical"
	alertmanagerConfig.Route.Routes[0].Receiver = "pagerduty"

	alertmanagerConfig.Route.Routes[1].GroupWait = "5m"
	alertmanagerConfig.Route.Routes[1].GroupInterval = "5m"
	alertmanagerConfig.Route.Routes[1].RepeatInterval = "5m"
	alertmanagerConfig.Route.Routes[1].Receiver = "pagerduty"
	alertmanagerConfig.Route.Routes[1].Match.Severity = "critical"
	alertmanagerConfig.Route.Routes[1].Receiver = "pagerduty"
	alertmanagerConfig.Route.Routes[1].Match.Severity = "critical"
	alertmanagerConfig.Route.Routes[1].Receiver = "deadmansnitch"
	
	alertmanagerConfig.Receivers[0].Name = "pagerduty"
	alertmanagerConfig.Receivers[0].PagerdutyConfigs =  []PagerDutyConfig{{}}
	alertmanagerConfig.Receivers[0].PagerdutyConfigs[0].ServiceKey = "pagerdutyServiceKey"
	
	alertmanagerConfig.Receivers[1].Name = "deadmansnitch"
	alertmanagerConfig.Receivers[1].WebhookConfigs = []WebhookConfig{{}}
	alertmanagerConfig.Receivers[1].WebhookConfigs[0].Url = "http://www.deadmansnitch.com"
```